src/sentry/api/endpoints/organization_member_index.py

```diff
@@ -43,3 +72,86 @@ def get(self, request, organization):
             on_results=lambda x: serialize(x, request.user),
             paginator_cls=OffsetPaginator,
         )
+
+    def post(self, request, organization):
+        """
+        Add a Member to Organization
+        ````````````````````````````
+
+        Invite a member to the organization.
+
+        :pparam string organization_slug: the slug of the organization the member will belong to
+        :param string email: the email address to invite
+        :param string role: the role of the new member
+        :param array teams: the slugs of the teams the member should belong to.
+
+        :auth: required
+        """
+        # TODO: If the member already exists, should this still update the role and team?
+        # For now, it doesn't, but simply returns the existing object
+
+        serializer = OrganizationMemberSerializer(data=request.DATA)
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+
+        result = serializer.object
+
+        _, allowed_roles = get_allowed_roles(request, organization)
+
+        # ensure listed teams are real teams
+        teams = list(Team.objects.filter(
+            organization=organization,
+            status=TeamStatus.VISIBLE,
+            slug__in=result['teams'],
+        ))
+
+        if len(set(result['teams'])) != len(teams):
+            return Response({'teams': 'Invalid team'}, 400)
+
+        if not result['role'] in {r.id for r in allowed_roles}:
+            return Response({'role': 'You do not have permission to invite that role.'}, 403)
+
+        # This is needed because `email` field is case sensitive, but from a user perspective,
+        # Sentry treats email as case-insensitive (Eric@sentry.io equals eric@sentry.io).
+        try:
+            OrganizationMember.objects.filter(
+                organization=organization,
+                user__email__iexact=result['email'],
+                user__is_active=True,
+            )[0]
+
+        except IndexError:
+            pass
+        else:
+            return Response(
+                {'email': 'The user %s is already a member' % result['email']}, 409)
+
+        om = OrganizationMember(
+            organization=organization,
+            email=result['email'],
+            role=result['role'])
+
+        if settings.SENTRY_ENABLE_INVITES:
+            om.token = om.generate_token()
+
+        try:
+            with transaction.atomic():
+                om.save()
+        except IntegrityError:
+            return Response({'email': 'The user %s is already a member' % result['email']}, 409)
+
+        self.save_team_assignments(om, teams)
```

Comment:
In theory, this can fail due to a race condition with an IntegrityError. Just pointing it out. I'm not sure it'll actually be a problem in practice, but it's absolutely possible for this to error and spit back a 500. But we'll see this in Sentry.

Reply:
hm, what would you suggest if you were more worried about this? a retry?

Reply:
Or a lock to prevent it entirely.

Reply:
https://github.com/getsentry/sentry/blob/master/src/sentry/models/project.py#L136-L142
like this?